katdal/chunkstore.py

```diff
@@ -77,32 +92,43 @@ def generate_chunks(shape, dtype, max_chunk_size, dims_to_split=None,
     """
     if dims_to_split is None:
         dims_to_split = range(len(shape))
-    dataset_size = np.prod(shape) * np.dtype(dtype).itemsize
-    # The ideal number of chunks to achieve requested chunk size (can be float)
-    num_chunks = dataset_size / max_chunk_size
-    # Start with the whole array as a single big chunk
-    chunks = [(s,) for s in shape]
+    if max_dim_elements is None:
+        max_dim_elements = {}
+
+    dim_elements = list(shape)
+    for i in dims_to_split:
+        if i in max_dim_elements and max_dim_elements[i] < shape[i]:
+            if power_of_two:
+                dim_elements[i] = _floor_power_of_two(max_dim_elements[i])
+            else:
+                dim_elements[i] = max_dim_elements[i]
+    cur_elements = int(np.prod(dim_elements))
+    # The ideal number of elements per chunk to achieve requested chunk size
+    # (can be float).
+    max_elements = max_chunk_size / np.dtype(dtype).itemsize
     # Split the array greedily along each dimension, in order of `dims_to_split`
     for dim in dims_to_split:
+        if cur_elements <= max_elements:
+            break      # We have already split enough to meet the budget
         if dim >= len(shape):
```

Comment:
This is not expected, right? Should you not rather do an `IndexError` or similar? Unless `dims_to_split` is merely a suggestion.

Reply:
I was just leaving in the semantics you wrote in the first place. But I'm happy to remove it. I don't think it's necessary to explicitly check it, since it's lead to an IndexError anyway.